src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs

```diff
@@ -519,68 +537,133 @@ private static bool ShouldOptimizeOutInitializer(BoundStatement initializer)
+        //        }
+        //
+        //        async static Task<string> M1()
+        //        {
+        //            x = 42;
+        //            await Task.Yield();
+        //            return "";
+        //        }
+        //
+        // Example2:
+        //        static int x = 123;
+        //        static string Test1()
+        //        {
+        //            // cannot replace value of "x" with a reference to "x"
+        //            // since that would make the method see the mutations in M1();
+        //            return (x + 0).ToString(M1());
+        //        }
+        //
+        //        static string M1()
+        //        {
+        //            x = 42;
+        //            return "";
+        //        }
+        //
+        // #2: Ordinary byval argument passed to an "in" parameter.
+        // 
+        // The language only requires that ordinary byval arguments must be readable (RValues are ok).
+        // However if the target parameter is an "in" parameter, the underlying implementation passes by reference.
+        //
+        // Example:
+        //        static int x = 123;
+        //        static void Main(string[] args)
+        //        {
+        //            // cannot replace value of "x" with a direct reference to x
+        //            // since Test will see unexpected changes due to aliasing.
+        //            Test(x + 0);
+        //        }
+        //
+        //        static void Test(in int y)
+        //        {
+        //            Console.WriteLine(y);
+        //            x = 42;
+        //            Console.WriteLine(y);
+        //        }
+        //
+        // NB: The readonliness is not considered here. 
+        //     We only care about possible introduction of aliasing. I.E. RValue->LValue change.
+        //     Even if we start with a readonly variable, it cannot be lowered into a writeable one,
+        //     with one exception - spilling of the value into a local, which is ok.
+        //
+        internal static bool CanBePassedByReference(BoundExpression expr)
+        {
+            if (expr.ConstantValue != null)
             {
                 return false;
             }
 
-            switch (receiver.Kind)
+            switch (expr.Kind)
             {
                 case BoundKind.Parameter:
                 case BoundKind.Local:
                 case BoundKind.ArrayAccess:
                 case BoundKind.ThisReference:
-                case BoundKind.BaseReference:
                 case BoundKind.PointerIndirectionOperator:
+                case BoundKind.PointerElementAccess:
                 case BoundKind.RefValueOperator:
                 case BoundKind.PseudoVariable:
+                case BoundKind.DiscardExpression:
+                    return true;
+
+                case BoundKind.DeconstructValuePlaceholder:
+                    // is this always a proxy for a temp local?
```

Comment:
Here's the link to replace the comment: [https://github.com/dotnet/roslyn/issues/24160](https://github.com/dotnet/roslyn/issues/24160)